src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,340 @@
+        if hash_cache.as_slice() == hash_on_chain {
+            cache_hit = true;
+            trace!(logger, "The canister wasm was found in the cache.");
+            let wasm_url_str = std::fs::read_to_string(&wasm_url_txt_path)
+                .with_context(|| format!("Failed to read {:?}", &wasm_url_txt_path))?;
+            pulled_canister.wasm_url = Some(wasm_url_str);
+        }
+    }
+    if !cache_hit {
+        // fetch `dfx:wasm_url`
+        let wasm_url_raw = fetch_metatdata(agent, canister_id, DFX_WASM_URL)
+            .await?
+            .ok_or_else(|| {
+                anyhow!("`{DFX_WASM_URL}` metadata not found in canister {canister_id}.")
+            })?;
+        let wasm_url_str = String::from_utf8(wasm_url_raw)?;
+        let wasm_url = reqwest::Url::parse(&wasm_url_str)?;
+        write_to_tempfile_then_rename(wasm_url_str.as_bytes(), &wasm_url_txt_path)?;
+        pulled_canister.wasm_url = Some(wasm_url_str);
+
+        // download
+        let response = reqwest::get(wasm_url.clone()).await?;
+        let status = response.status();
+        if status.is_client_error() || status.is_server_error() {
+            bail!("Failed to download wasm from url: {wasm_url}.\n  StatusCode: {status}");
+        }
+        let content = response.bytes().await?;
+
+        // hash check
+        let mut hasher = Sha256::new();
+        hasher.update(&content);
+        let hash_download = hasher.finalize();
+        if hash_download.as_slice() != hash_on_chain {
+            bail!(
+                "Hash mismatch.
+on chain: {}
+download: {}",
+                hex::encode(hash_on_chain),
+                hex::encode(hash_download.as_slice())
+            );
+        }
+
+        write_to_tempfile_then_rename(&content, &wasm_path)?;
+    }
+    // get `candid:service` from downloaded wasm
+    let wasm = std::fs::read(&wasm_path)
+        .with_context(|| format!("Failed to read wasm from {:?}", &wasm_path))?;
+    let candid_service = get_metadata(&wasm, CANDID_SERVICE).with_context(|| {
+        format!(
+            "Failed to get {} metadata from {:?}",
+            CANDID_SERVICE, &wasm_path
+        )
+    })?;
+    let service_candid_path = get_service_candid_path(canister_id)?;
+    write_to_tempfile_then_rename(candid_service.as_bytes(), &service_candid_path)?;
+
+    // get `candid:args` from downloaded wasm
+    let candid_args = get_metadata(&wasm, CANDID_ARGS).with_context(|| {
+        format!(
+            "Failed to get {} metadata from {:?}",
+            CANDID_ARGS, &wasm_path
+        )
+    })?;
+    pulled_canister.candid_args = Some(candid_args);
+
+    // try get `dfx:init` from downloaded wasm
+    match get_metadata(&wasm, DFX_INIT) {
+        Ok(dfx_init) => pulled_canister.dfx_init = Some(dfx_init),
+        Err(_e) => trace!(
+            logger,
+            "Failed to get {} metadata from {:?}",
+            DFX_INIT,
+            &wasm_path
+        ),
+    }
```

Comment:
Suggestion: update to a newer version of `ic-wasm`, which changes this method to operate on a Module rather than a wasm blob, and return an `Option<String>` rather than a `Result<String>`.

Reply:
I would like to make the change once this PR https://github.com/dfinity/sdk/pull/3090 merged.